text/0000-variadic-tuples.md

```diff
@@ -0,0 +1,1006 @@
+fn my_func<Head, (..T)>((head, v @ ..): (Head, ..T)) -> (..T) { 
+    ...
+}
+
+impl<Head, (..T)> Clone for (Head, ..T) 
+where 
+    ..(T: Clone),
+    Head: Clone, {
+  fn clone(&self) -> Self {
+    // We destructure `*self` which has a variadic tuple type `(Head, ..T)`
+    let (ref head, ref v @ ..) = *self;
+    ...
+  }
+}
+```
+
+### Iterating over variadic tuple
+
+We can iterate over the member of a variadic tuple or over the type of a variadic tuple type.
+
+*Important note: the iteration is inlined by the compiler, it is not a generic runtime heterogenous iteration of tuple members.*
+
+We use the following syntax to iterate on variadic tuples:
+
+```rust
+// The result of the for block is a variadic tuple made of
+// the result of each iteration
+let result: (..Option<&V>) = {
+    // `key` and `map` are variables iterating the variadic tuples `k: (..K)` and `maps: (..&HashMap<K, V>)`, `key` will iterate by reference (because of the ref keyword)
+    // `KEY` and `VALUE` are type variables iterating the variadic tuple types `(..K)` and `(..V)`
+    // `(k, maps)` declares the iterated variadic tuples `k` and `maps`
+    // `<K, V>` declares the iterated variadic tuple types
+    (for (ref key, map) <KEY, VALUE> @in (k, maps) <K, V> {
+        HashMap::<KEY, VALUE>::get(&map, key)
+    })
+};
+```
+
+Note: when iterating over multiple variadic tuple or variadic tuple types, they must have all the same arity. To ensure this, all variadic tuple types involved must have been declared together.
+
+Examples:
+
+```rust
+impl<(..(K, V))> MegaMap<(..(K, V))>
+where ..(K: Hash), {
+    fn get(&self, k: (..K)) -> (..Option<V>) {
+        let result: (..Option<&V>) = {
+            (for (ref k, map) <Key, Value> @in (k, &self.maps) <K, V> {
+                HashMap::<Key, Value>::get(&map, k)
+            })
+        };
+
+        result
+    }
+}
+
+impl<(..T), Last> Hash for (..T, Last)
+where
+    ..(T: Hash),
+    Last: Hash + ?Sized, {
+
+    #[allow(non_snake_case)]
+    fn hash<S: Hasher>(&self, state: &mut S) {
+        let (ref tuple @ .., ref last) = *self;
+       
+        // Use case: only variadic tuple
+        for member @in tuple {
+          member.hash(state);
+        };
+        last.hash(state);
+
+        // Use case: variadic tuple and type
+        for member <H> @in tuple <T> {
+          <T as Hash>::hash(&member, state);
+        };
+        last.hash(state);
+    }
+}
+
+trait Merge<(..R)> {
+    type Value;
+    fn merge(self, r: (..R)) -> Self::Value;
+}
+
+impl<(..L), (..R)> Merge<(..R)> for (..L) {
+    type Value = (..L, ..R);
+
+    fn merge(self, r: (..R)) -> Self::Value {
+        (
+            for l1 @in self { l1 },
+            for r1 @in r { r1 },
+        )
+    }
+}
+
+trait Integer {
+    fn one() -> Self;
+}
+
+fn add_one<(..T)>((..t): (..T)) -> (..T)
+where
+    ..(T: Integer + Add), {
+    (for t1 <T1> @in t <T> { t1 + T1::one() })
+}
```

Comment:
I really just don't like this syntax. I think that it's 100% fine to leave the initial implementation recursion-only until an iterative syntax is decided upon.

I do agree that an iterative syntax is a good idea, but a `for @in` loop feels kinda hacky and gross imho, which is the reason why we haven't accepted a variadic tuple syntax yet.